### PR TITLE
add retryConnect() when using session pool

### PR DESCRIPTION
**Pull request by songqing**
## What type of PR is this?
- [x] bug
- [ ] feature
- [ ] enhancement

#### Description:
When using SessionPool to query, there will be some connect error as the connection could be broken, then, using these connection to query will fail.

## How do you solve it?
Add retryConnect() when query failed.
This implementation refers to [the java client](https://github.com/vesoft-inc/nebula-java/blob/b703e03bb062b8fac8989ddc35f2b3243b4d3510/client/src/main/java/com/vesoft/nebula/client/graph/net/Session.java#L116)

### CODE REVIEW

**Comment by songqing** [comment]
@Shylock-Hg @jievince Could you help to review?

File: src/client/Session.cpp [hunk 1]
```diff
@@ -22,7 +23,37 @@ void Session::asyncExecute(const std::string &stmt, ExecuteCallback cb) {
 
 ExecutionResponse Session::executeWithParameter(
     const std::string &stmt, const std::unordered_map<std::string, Value> &parameters) {
-  return ExecutionResponse(conn_.executeWithParameter(sessionId_, stmt, parameters));
+  if (connectionIsBroken_ && retryConnect_) {
```
Comment (Shylock-Hg) [inline src/client/Session.cpp:26]: Why needs this block?

Comment (songqing) [reply src/client/Session.cpp:26]: This implementation is from [java client](https://github.com/vesoft-inc/nebula-java/blob/b703e03bb062b8fac8989ddc35f2b3243b4d3510/client/src/main/java/com/vesoft/nebula/client/graph/net/Session.java#L116), when doing a query, the connection maybe broken, so, we can retry the connection automatically to make sure the query can be successful.

Comment (Shylock-Hg) [reply src/client/Session.cpp:26]: Got.

**Review by Shylock-Hg — DISMISSED** [review]

**Comment by songqing** [comment]
@Shylock-Hg UT failed, could you help review again?
f2

**Review by Shylock-Hg — APPROVED** [review]

**Comment by wey-gu** [comment]
Thanks a lot @songqing for the contribution, and welcome to the NebulaGraph community!

PS, given that you're a DGL contributor, I'd like to share my slow project(based on nebula-python, slow in data loading: https://github.com/wey-gu/nebula-dgl ), as I’m just curious, are you working on something that's related to both dgl and NebulaGrpah?

@QingZ11 we have a new contributor now!

**Comment by QingZ11** [comment]
> Thanks a lot @songqing for the contribution, and welcome to the NebulaGraph community!
> 
> PS, given that you're a DGL contributor, I'd like to share my slow project(based on nebula-python, slow in data loading: https://github.com/wey-gu/nebula-dgl ), as I’m just curious, are you working on something that's related to both dgl and NebulaGrpah?
> 
> @QingZ11 we have a new contributor now!

😂 actually，[songqing](https://github.com/songqing)'s first merged pr is this one [#5618](https://github.com/vesoft-inc/nebula/pull/5618).

@songqing Hi, you have a contributor gift waiting to be shipped. If it's convenient for you, could you please provide your shipping information? link: https://wj.qq.com/s2/7782511/2cef/

Apologies for not reaching out to you in a timely manner. I discovered your contribution in the release note for [v3.6.0](https://discuss.nebula-graph.com.cn/t/topic/13771) and added your GitHub avatar at the end of the release note. Once again, I apologize for not contacting you earlier.

他第一个 pr 不是这个啦，是在刚发布的 v3.6.0 里的 [#5618](https://github.com/vesoft-inc/nebula/pull/5618)。因为一些事情耽误了，本来想稍后联系下 songqing 的。cc @wey-gu 

@songqing 你有一份待领取的贡献者礼物，如果你方便的话，可以填写下收件信息么？是这个链接哈：https://wj.qq.com/s2/7782511/2cef/ （如果这个问卷有什么需要修改的地方，也可以在问卷最后备注哈，谢谢啦）

抱歉，没能及时联系你，在 v3.6.0 的 release note 中我发现了你贡献的身影，在 release note 末加上了你的 GitHub 头像哈。再次抱歉，没能及时联系你

**Comment by songqing** [comment]
> 

Thank you~ @wey-gu, now, I do some research on Nebula and DGL both, but using them separately, I'm interested with your repo, will have a good look it later.

**Comment by songqing** [comment]
> 

Thank you @QingZ11 , I'm glad to contribute as I have learned a lot from our code, that's already a gift~

**Comment by QingZ11** [comment]
You have already got a personalized contributor certificate (physical) from NebulaGraph, which one is like the certificate as below.

image

BTW, we will also sent you an electronic certificate via email. The certificate numbering is currently being processed through a script, so the email is expected to be sent to your inbox around Wednesday.

**Comment by songqing** [comment]
> 

OK, thank you~

**Comment by wey-gu** [comment]
> Thank you~ @wey-gu, now, I do some research on Nebula and DGL both, but using them separately, I'm interested with your repo, will have a good look it later.

Wow, awesome! @songqing

> 
> Thank you @QingZ11 , I'm glad to contribute as I have learned a lot from our code, that's already a gift~

It's so lucky to have you in NebulaGraph Community.

**Pull request merged**